weasyl/test/test_two_factor_auth.py

```diff
@@ -0,0 +1,221 @@
+from __future__ import absolute_import, unicode_literals
+
+import re
+import urllib
+
+import pyotp
+import pytest
+from qrcodegen import QrCode
+
+from libweasyl import security
+from weasyl import define as d
+from weasyl import two_factor_auth as tfa
+from weasyl.test import db_utils
+
+
+@pytest.mark.usefixtures('db')
+def test_get_number_of_recovery_codes():
+    user_id = db_utils.create_user()
+    
+    # This /should/ be zero right now, but verify this for test environment sanity.
+    assert 0 == d.engine.scalar("""
+        SELECT COUNT(*)
+        FROM twofa_recovery_codes
+        WHERE userid = (%(userid)s)
+    """, userid=user_id)
+    assert tfa.get_number_of_recovery_codes(user_id) == 0
+    d.engine.execute("""
+        INSERT INTO twofa_recovery_codes (userid, recovery_code)
+        VALUES ( (%(userid)s), (%(code)s) )
+    """, userid=user_id, code=security.generate_key(20))
+    assert tfa.get_number_of_recovery_codes(user_id) == 1
+    d.engine.execute("""
+        DELETE FROM twofa_recovery_codes
+        WHERE userid = (%(userid)s)
+    """, userid=user_id)
+    assert tfa.get_number_of_recovery_codes(user_id) == 0
+
+
+@pytest.mark.usefixtures('db')
+def test_generate_recovery_codes():
+    user_id = db_utils.create_user()
+
+    recovery_codes = tfa.generate_recovery_codes(user_id)
+    assert len(recovery_codes) == 10
+
+    query = d.engine.execute("""
+        SELECT recovery_code
+        FROM twofa_recovery_codes
+        WHERE userid = (%(userid)s)
+    """, userid=user_id).fetchall()
+    for code in query:
+        assert code['recovery_code'] in recovery_codes
```

Comment:
Add above; for sanity. And because it is good to test our assumptions.
``assert len(code['recovery_code']) == 20``